benchmarks/linalg_benchmark.cpp

```diff
@@ -0,0 +1,71 @@
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/GPUVector.h>
+#include <shogun/mathematics/linalg/linalg.h>
+#include <vector>
+#include <algorithm>
+#include <hayai/hayai.hpp>
+
+using namespace shogun;
+
+/**
+ * Instructions :
+ * 1. Install benchmarking toolkit "hayai" (https://github.com/nickbruun/hayai)
+ * 2. Compile against libhayai_main, e.g.
+ * 		g++ -O3 -std=c++11 matrix_product_benchmark.cpp -I/usr/include/eigen3 \
+ *		-lshogun -lhayai_main -lOpenCL -o benchmark
+ * 3. ./benchmark
+ */
+
+/** Generate data only once */
+struct Data
+{
+	Data()
+	{
+		A = init(1);
+		B = init(2);
+		Av = init(A);
+		Bv = init(B);
+	}
+
+	SGVector<int32_t> init(int32_t begin)
+	{
+		index_t n = num_rows;
+		std::vector<int32_t> mem(n);
+		std::iota(mem.data(), mem.data() + n, begin);
+
+		SGVector<int32_t> m(num_rows);
+		memcpy(m.vector, mem.data(), n*sizeof(int32_t));
+
+		return m;
+	}
+
+	CGPUVector<int32_t> init(SGVector<int32_t> m)
+	{
+		CGPUVector<int32_t> mv(m);
+		return mv;
+	}
+
+	SGVector<int32_t> A;
+	SGVector<int32_t> B;
+	CGPUVector<int32_t> Av;
+	CGPUVector<int32_t> Bv;
+
+	static constexpr index_t num_rows = 100;
+};
+
+BENCHMARK_P(SGVector, dot_product_eigen3, 10, 1000,
+		(SGVector<int32_t> A, SGVector<int32_t> B))
+{
+	auto C = linalg::dot<linalg::Backend::EIGEN3>(A, B);
```

Comment:
Oh I see. You wanted to compare between the old code and the new one. I think this is not required.